src/main/java/com/arangodb/springframework/core/util/ArangoExceptionTranslator.java

```diff
@@ -20,63 +20,76 @@
 
 	@Override
+	@Nullable
 	public DataAccessException translateExceptionIfPossible(final RuntimeException ex) {
-		DataAccessException dae = null;
-		if (ex instanceof DataAccessException) {
-			dae = DataAccessException.class.cast(ex);
-		} else if (ex instanceof ArangoDBException) {
-			final ArangoDBException exception = ArangoDBException.class.cast(ex);
+		if (ex instanceof DataAccessException exception) {
+			return exception;
+		}
+		if (ex instanceof ArangoDBException exception) {
 			final Integer responseCode = exception.getResponseCode();
-			if (responseCode != null) {
-				switch (responseCode) {
-				case ArangoErrors.ERROR_HTTP_UNAUTHORIZED:
-				case ArangoErrors.ERROR_HTTP_FORBIDDEN:
-					dae = new PermissionDeniedDataAccessException(exception.getMessage(), exception);
-					break;
-				case ArangoErrors.ERROR_HTTP_BAD_PARAMETER:
-				case ArangoErrors.ERROR_HTTP_METHOD_NOT_ALLOWED:
-					dae = new InvalidDataAccessApiUsageException(exception.getMessage(), exception);
-					break;
-				case ArangoErrors.ERROR_HTTP_NOT_FOUND:
-					dae = new InvalidDataAccessResourceUsageException(exception.getMessage(), exception);
-					break;
-				case ArangoErrors.ERROR_HTTP_CONFLICT:
-					dae = new DataIntegrityViolationException(exception.getMessage(), exception);
-					break;
-				case ArangoErrors.ERROR_HTTP_PRECONDITION_FAILED:
-				case ArangoErrors.ERROR_HTTP_SERVICE_UNAVAILABLE:
-					dae = new DataAccessResourceFailureException(exception.getMessage(), exception);
-					break;
-				case ArangoErrors.ERROR_HTTP_SERVER_ERROR:
-				default:
-					dae = new ArangoUncategorizedException(exception.getMessage(), exception);
-					break;
-				}
+			if (responseCode == null) {
+				return new ArangoUncategorizedException(exception.getMessage(), exception);
 			}
+			BiFunction<String, ArangoDBException, DataAccessException> constructor = switch (responseCode) {
+				case ERROR_HTTP_UNAUTHORIZED, ERROR_HTTP_FORBIDDEN -> PermissionDeniedDataAccessException::new;
+				case ERROR_HTTP_BAD_PARAMETER, ERROR_HTTP_METHOD_NOT_ALLOWED -> InvalidDataAccessApiUsageException::new;
+				case ERROR_HTTP_NOT_FOUND -> mostSpecific(exception, Map.ofEntries(
+							entry(hasErrorNumber(ERROR_ARANGO_DOCUMENT_NOT_FOUND), DataRetrievalFailureException::new)
+						), InvalidDataAccessResourceUsageException::new);
+				case ERROR_HTTP_CONFLICT -> mostSpecific(exception, Map.ofEntries(
+							entry(hasErrorNumber(ERROR_ARANGO_CONFLICT).and(errorMessageContains("write-write")), TransientDataAccessResourceException::new),
+							entry(hasErrorNumber(ERROR_ARANGO_CONFLICT).and(errorMessageContains("_rev")), OptimisticLockingFailureException::new),
+							entry(hasErrorNumber(ERROR_ARANGO_UNIQUE_CONSTRAINT_VIOLATED).and(errorMessageContains("_key")), DuplicateKeyException::new)
+						),
+						DataIntegrityViolationException::new);
+				case ERROR_HTTP_PRECONDITION_FAILED -> OptimisticLockingFailureException::new;

```

Comment:
```suggestion
				case ERROR_HTTP_PRECONDITION_FAILED -> mostSpecific(exception, Map.ofEntries(
								entry(hasErrorNumber(ERROR_ARANGO_CONFLICT).and(errorMessageContains("_rev")), OptimisticLockingFailureException::new)
						),
						DataAccessResourceFailureException::new);
```

Only `code: 412, errorNum: 1200` should be considered `OptimisticLockingFailureException`. I.e. the db could also return `code: 412, errorNum: 21004` in case of queue time violation, see https://www.arangodb.com/docs/stable/http/general.html